doc/lsst.daf.butler/writing-subcommands.rst

```diff
@@ -0,0 +1,320 @@
+
+Uses a shared option:
+
+.. code-block:: py
+   :name: shared-argument-use
+
+   @click.command()
+   @example_argument()
+   def cmd(example):
+       ...
+
+Butler Command Plugins
+======================
+
+Packages can add subcommands to the butler command using a plugin system. This section describes how to do
+that. To use the plugin system you should also read and understand the sections above about `the butler
+command`_. Then, write your subcommands and arrange them as described below in `Package Layout`_. Finally,
+declare them as ``butler`` command plugins as described in `Manifest`_.
+
+Package Layout
+--------------
+
+- All command line interface code should go in a folder called ``cli`` under the package's python
+  hierarchy e.g. ``python/lsst/daf/butler/cli``.
+- Commands go in a file ``.../cli/cmd/commands.py``
+- Options go in a file ``.../cli/opt/options.py``
+- Shared options go in a file ``.../cli/opt/sharedOptions.py``
+- Arguments go in a file ``.../cli/opt/arguments.py``
+- To make commands importable there must be a manifest file, usually named ``resources.yaml`` in the ``cli``
+  folder. ``cli``
+- There may be a ``utils.py`` file, also usually in the ``cli`` folder.
+
+.. code-block:: text
+
+   cli
+   ├── cmd
+   │    └── commands.py
+   ├── opt
+   │    ├── arguments.py
+   │    ├── options.py
+   │    └── sharedOptions.py
+   ├── resources.yaml
+   └── utils.yaml
+
+Manifest
+--------
+
+The ``butler`` command finds plugin commands by way of a resource manifest published in an environment
+variable.
+
+Create a file ``resources.yaml`` in the ``cli`` folder. ``cmd`` is the section for importable commands;
+``import`` names the package that the commands can be imported from and ``commands`` is a list of importable
+command names. Use the dash-separated command name, not the underscore-separated function name.
+
+For example, the manifest file for ``butler`` plugin subcommands in ``obs_base`` is like this:
+
+.. code-block:: yaml
+
+    cmd:
+      import: lsst.obs.base.cli.cmd
+      commands:
+        - register-instrument
+        - write-curated-calibrations
+
+Publish the resource manifest in an environment variable: in the package's ``ups/<pkg>.table`` file, add a
+command to prepend ``DAF_BUTLER_PLUGINS`` with the location of the resource manifest. Make sure to use the
+environment variable for the location of the package.
+
+The settings for ``obs_base`` are like this:
+
+.. code-block:: text
+
+    envPrepend(DAF_BUTLER_PLUGINS, $OBS_BASE_DIR/python/lsst/obs/base/cli/resources.yaml)
+
+Command Loader Base Class
+=========================
+
+The ``butler`` commmand uses a ``click.MultiCommand`` subclass called ``LoaderCLI`` that dynamically loads
+subcomands from the local package and from plugin packages. ``LoaderCLI`` can be used to implement other
+commands that dynamically load subcommands.
+
+It's easy to create a new kind of command by copying the template below and making a few small changes:
+
```

Comment:
If there is click framework code that can be used by commands that don't use a butler (or specifically butler command line options), then shouldn't the code be moved outside daf_butler?

Reply:
I spoke with @timj about this. It would be better in a different package, but right now we want to be removing dependencies from daf_butler (as opposed to adding them), and because having it here is not a gen2 deprecation blocker, we're not going to do it right now.